zk_coordinator.go

```diff
@@ -588,132 +588,96 @@ func (this *ZookeeperCoordinator) tryRemoveOldApiRequests(group string, api Cons
+			return deadline, nil
 		}
-
-		if err == nil {
-			go func() {
-				for {
-					barrierTimeout := deadline.Sub(time.Now())
-					select {
-					case e, ok := <-zkMemberJoinedWatcher:
-						{
-							Debugf(this, "Member joined channel %v", e)
-							if ok && e.Type != zk.EventNotWatching && e.State != zk.StateDisconnected {
-								ask := make(chan bool)
-								memberJoinedWatcher <- ask
-								if <-ask {
-									return
-								}
-							} else {
-								_, _, zkMemberJoinedWatcher, err = this.zkConn.ChildrenW(path)
-								Debugf(this, "Trying to renew watcher at %s", path)
-							}
-						}
-					case <-time.After(barrierTimeout):
-						{
-							Infof(this, "Barrier timed out after %v", barrierTimeout)
-							memberJoinedWatcher <- nil
-							return
-						}
-					}
-				}
-			}()
-
-			Infof(this, "Successfully joined state barrier %s", path)
-			return memberJoinedWatcher, err
-		}
-
-		Warnf(this, "Failed to join state barrier %s, retrying...", path)
+		Warnf(this, "Failed to join state barrier %s, retrying...", barrierPath)
 	}
-
-	Errorf(this, "Failed to join state barrier %s after %d retries", path, this.config.MaxRequestRetries)
-
-	return nil, err
+	return time.Now(), fmt.Errorf("Failed to join state barrier %s after %d retries", barrierPath, this.config.MaxRequestRetries)
 }
 
-func (this *ZookeeperCoordinator) isStateBarrierPassed(group string, stateHash string, api string, expected int) (bool, error) {
-	Debugf(this, "Trying to assert rebalance state for group %s and hash %s with %d", group, stateHash, expected)
-	path := fmt.Sprintf("%s/%s/%s", newZKGroupDirs(this.config.Root, group).ConsumerApiDir, api, stateHash)
-	var children []string
-	var err error
-	backoffMultiplier := 1
-	for i := 0; i <= this.config.MaxRequestRetries; i++ {
-		children, _, err = this.zkConn.Children(path)
-		if err == zk.ErrNoNode {
-			return false, err
-		} else if err == nil {
-			return len(children) == expected, err
+func (this *ZookeeperCoordinator) waitForMembersToJoin(barrierPath string, expected int, doneChan chan <- error, stopChan <- chan struct{}) {
+	// Make sure we clean up the channel.
+	defer close(doneChan)
+
+	for {
+		select {
+		// Using a priority select to provide precedence to the stop chan
+		case <-stopChan:
+			return
+		default:
+			children, _, zkMemberJoinedWatcher, err := this.zkConn.ChildrenW(barrierPath)
+			if err != nil && err == zk.ErrNoNode {
+				doneChan <- fmt.Errorf("%v; path: %s", err, barrierPath)
+				return
+			} else if len(children) == expected {
+				doneChan <- nil
+				// don't leave the zkMemberJoinedWatcher chan out there with no one to receive the message it produces later as it would cause a block.
+				go func(blackhole <-chan zk.Event) {
+					<- blackhole
+				}(zkMemberJoinedWatcher)
+				return
+			}
+			// Haven't seen all expected consumers on this barrier path.  Watch for changes to the path...
+			select {
+			case <-zkMemberJoinedWatcher:
+				continue
+			}
```

Comment:
Shouldn't this select also have a case <-stopChan? It looks it might block here indefinitely